impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,200 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
-                        
-                        Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
-                        
-                        if (groupMembers.hasMoreElements()) {
-                            return groupMembers.nextElement();
-                        }
-                    }
-                    break;
-                case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
-                    try {
-                        return 
-                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
-
+            // If the principal is a standard one, return it
             if (CallerPrincipal.class.isAssignableFrom(principal.getClass())) {
                 return principal;
             }
-            
+
+            // Do some checks to determine it from vendor specific data
+            Principal vendorCallerPrincipal = getVendorCallerPrincipal(principal);
+            if (vendorCallerPrincipal != null) {
+                return vendorCallerPrincipal;
+            }
+
+            // Check for Servlet
             try {
                 return CDI.current().select(HttpServletRequest.class).get().getUserPrincipal();
             } catch (Exception e) {
                 // Not inside an HttpServletRequest
             }
-         
+
+            // Check for EJB
             EJBContext ejbContext = EJB.getEJBContext();
             if (ejbContext != null) {
-                String principalName = ejbContext.getCallerPrincipal().getName();
-             
-                // EJBs return empty/null name for unauthenticated callers
-                return principalName != null && !principalName.isEmpty() ? ejbContext.getCallerPrincipal() : null;
+                // The EJB returned value must be verified for its "authenticated name" since it's vendor specific
+                return getVendorCallerPrincipal(ejbContext.getCallerPrincipal());
             }
 
             return null;
         }
-        
+
+        return null;
+    }
+
+    /**
+     * Get the underlying caller principal based on vendor specific (e.g.: class
+     * names, EJB unauthenticated name, etc)
+     *
+     * @param principal
+     * @return
+     */
+    private Principal getVendorCallerPrincipal(Principal principal) {
```

Comment:
It's too much work to refactor now, but I'm not fond of this big case statement, since it assumes the given types will only ever appear in a Subject if they're the caller principal, and that there will only ever be one of them. See #71 .